### PR TITLE
Migrate 'components/post_emoji' module and associated tests to TypeScript

### PR DESCRIPTION
**Pull request by AGMETEOR**
**Summary**

MM-18968-Migrate 'components/post_emoji' module and associated tests to TypeScript

**Ticket Link**

[https://mattermost.atlassian.net/browse/MM-18968](https://mattermost.atlassian.net/browse/MM-18968)
Fixes https://github.com/mattermost/mattermost-server/issues/12460

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -5167,6 +5167,12 @@
           "resolved": "https://registry.npmjs.org/regexp-tree/-/regexp-tree-0.1.13.tgz",
           "integrity": "sha512-hwdV/GQY5F8ReLZWO+W1SRoN5YfpOKY6852+tBFcma72DKBIcHjPRIlIvQN35bCOljuAfP2G2iB0FC/w236mUw==",
           "dev": true
+        },
+        "semver": {
```
Comment (devinbinnie) [inline package-lock.json]: Can you revert your `package-lock.json` changes? They are unnecessary for the TS migration.

File: components/post_emoji/post_emoji.tsx [hunk 2]
```diff
@@ -1,11 +1,24 @@
...
+import React, {CSSProperties} from 'react';
 import PropTypes from 'prop-types';
 
-export default class PostEmoji extends React.PureComponent {
-    static propTypes = {
+interface PostEmojiProps {
+    name: string;
+    imageUrl: string;
+}
+declare module 'react' {
```
Comment (devinbinnie) [inline components/post_emoji/post_emoji.tsx]: Just curious, how come we are redefining HTML attributes within the `react` module?

File: components/post_emoji/post_emoji.tsx [hunk 3]
```diff
@@ -1,11 +1,24 @@
...
+    interface HTMLAttributes<T> extends AriaAttributes, DOMAttributes<T> {
+        alt?: string;
+        className?: string;
+        title?: string;
+        style?: CSSProperties;
+    }
+}
+
+export default class PostEmoji extends React.PureComponent<PostEmojiProps, {}> {
+    public static propTypes = {
```
Comment (devinbinnie) [inline components/post_emoji/post_emoji.tsx]: These are no longer needed now that we're using TypeScript.

Comment (AGMETEOR) [reply components/post_emoji/post_emoji.tsx]: @devinbinnie taken them out.

File: components/post_emoji/post_emoji.tsx (hunk 2 above)
Comment (AGMETEOR) [reply components/post_emoji/post_emoji.tsx]: @devinbinnie thanks for the review. Without redefining those, I was basically getting this error within the file. Looking at it closely now, it's the `alt` attribute on the `<span>` that causes the warning. 
```
Type '{ alt: string; className: string; title: string; style: { backgroundImage: string; }; }' is not assignable to type 'DetailedHTMLProps<HTMLAttributes<HTMLSpanElement>, HTMLSpanElement>'.
  Property 'alt' does not exist on type 'DetailedHTMLProps<HTMLAttributes<HTMLSpanElement>, HTMLSpanElement>'.
```

File: package-lock.json (hunk 1 above)
Comment (AGMETEOR) [reply package-lock.json]: Made the revert :)

File: components/post_emoji/post_emoji.tsx [hunk 4]
```diff
@@ -1,24 +1,21 @@
 // Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
 // See LICENSE.txt for license information.
 
-import React from 'react';
+import React, {CSSProperties} from 'react';
 import PropTypes from 'prop-types';
```
Comment (devinbinnie) [inline components/post_emoji/post_emoji.tsx]: @AGMETEOR you have the same `PropTypes` tests failing on this PR as in https://github.com/mattermost/mattermost-webapp/pull/3882. Try removing this line.

File: components/post_emoji/post_emoji.tsx (hunk 2 above)
Comment (devinbinnie) [reply components/post_emoji/post_emoji.tsx]: Ahh I see, `alt` only applies to `HTMLImageElement` type objects. I feel like there's some refactoring that needs to happen here, but we'll leave this for now.

**Pull request merged**